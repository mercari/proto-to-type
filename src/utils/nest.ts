import { flatten } from 'array-flatten';

export interface Nest {
  [key: string]: Nest;
}
export type NestHandler = (nest: Nest, name: string, depth: number) => void;

export const digNest = (root: Nest, names: string[], callback?: NestHandler): Nest => {
  let curr = root;
  for (let i = 0; i < names.length; i++) {
    const name = names[i];
    if (!(name in curr)) curr[name] = digNest.createNest();
    curr = curr[name];
    callback?.(curr, name, i);
  }
  return curr;
};
digNest.createNest = (): Nest => Object.create(null);

export type Layer = readonly [string, string]; // [header, footer]
export type LayerWithChildren = [string, LayerWithChildren[], string];
export type MidSerializer = (name: string, index: number) => Layer;
export type EndSerializer<T> = (entry: T, name: string, index: number, current: Layer) => Layer;

const createLayer = ([h, f]: Layer = ['', '']): LayerWithChildren => [h, [], f];

/**
 * Serialize entries with nest structure based on names split by dots.
 * @example
 * 'Japan.Tokyo.Minato' becomes like:
 * namespace Japan {
 *   namespace Tokyo {
 *     namespace Minato {
 *       // definitions
 *     }
 *   }
 * }
 * @param entries map contains entry and name
 * @param wrapper top level layer
 * @param midSerializer serializer for intermediate layer
 * @param endSerializer serializer for entry
 */
export const nestSerializer = <T>(
  entries: Map<T, string>,
  wrapper: Layer,
  midSerializer: MidSerializer,
  endSerializer: EndSerializer<T>,
) => {
  const rootNest = digNest.createNest();
  const rootLayer = createLayer(wrapper);
  const layers = new Map<Nest, LayerWithChildren>();
  layers.set(rootNest, rootLayer);

  // TODO: consider about name conflict
  // not only same names, 'Japan.Tokyo.Chiyoda' and 'Japan.Tokyo' also conflict
  entries.forEach((fullName, entry) => {
    const names = fullName.split('.').filter(Boolean);
    let parent = rootLayer;
    // `digNest` helps resolving fastly nest structure from names.
    // It ensures correct nest structure, so we can focus on construction in each layer and its parent.
    digNest(rootNest, names, (nest, name, index) => {
      if (layers.has(nest)) {
        parent = layers.get(nest)!;
      } else {
        const layer = createLayer(midSerializer(name, index));
        layers.set(nest, layer);
        parent[1].push(layer);
        parent = layer;
      }
    });
    const lastDepth = names.length - 1;
    const lastName = names[lastDepth];
    [parent[0], parent[2]] = endSerializer(entry, lastName, lastDepth, [parent[0], parent[2]]);
  });

  return flatten(rootLayer).join('');
};
