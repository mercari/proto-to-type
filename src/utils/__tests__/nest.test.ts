import { digNest, nestSerializer } from '../nest';
import prettier from 'prettier';

describe('digNest', () => {
  it('create nest structure by names', () => {
    const expected = {
      Japan: {
        Tokyo: {
          Minato: {},
          Chiyoda: {},
        },
        Osaka: {
          Umeda: {},
        },
      },
    };

    const root = digNest.createNest();
    digNest(root, ['Japan', 'Tokyo', 'Minato']);
    digNest(root, ['Japan', 'Tokyo', 'Chiyoda']);
    digNest(root, ['Japan', 'Osaka', 'Umeda']);

    expect(root).toEqual(expected);
  });

  it('calls callback in each level of the nest', () => {
    const root = digNest.createNest();
    digNest(root, ['Japan', 'Tokyo', 'Minato']);
    digNest(root, ['Japan', 'Tokyo', 'Chiyoda']);
    digNest(root, ['Japan', 'Osaka', 'Umeda']);

    const fn = jest.fn();

    digNest(root, ['Japan', 'Osaka', 'Umeda'], fn);
    expect(fn.mock.calls.length).toBe(3);

    expect(fn.mock.calls[0][0]).toBe(root.Japan);
    expect(fn.mock.calls[0][1]).toBe('Japan');
    expect(fn.mock.calls[0][2]).toBe(0);

    expect(fn.mock.calls[1][0]).toBe(root.Japan.Osaka);
    expect(fn.mock.calls[1][1]).toBe('Osaka');
    expect(fn.mock.calls[1][2]).toBe(1);

    expect(fn.mock.calls[2][0]).toBe(root.Japan.Osaka.Umeda);
    expect(fn.mock.calls[2][1]).toBe('Umeda');
    expect(fn.mock.calls[2][2]).toBe(2);
  });
});

describe('digNest.createNest', () => {
  it('returns empty object', () => {
    const nest = digNest.createNest();
    expect(nest).not.toHaveProperty('hasOwnProperty');
    expect(nest).not.toHaveProperty('constructor');
    expect(nest).not.toHaveProperty('__proto__');
  });
});

describe('nestSerializer', () => {
  it('works correctly', () => {
    const entries = new Map<readonly [boolean, boolean], string>();

    const Minato = [true, false] as const;
    const Chiyoda = [true, true] as const;
    const Umeda = [false, true] as const;
    entries.set(Minato, 'Japan.Tokyo.Minato');
    entries.set(Chiyoda, 'Japan.Tokyo.Chiyoda');
    entries.set(Umeda, 'Japan.Osaka.Umeda');

    const result = nestSerializer(
      entries,
      ['namespace Address {', '}'],
      (name) => [`namespace ${name} {`, '}'],
      ([inTokyo, endWithA], name) => [`type ${name} = { inTokyo: ${inTokyo}; endWithA: ${endWithA} `, '};'],
    );

    const formatted = prettier.format(result, { parser: 'typescript' });
    expect(result).toMatchSnapshot();
    expect(formatted).toMatchSnapshot();
  });
});
