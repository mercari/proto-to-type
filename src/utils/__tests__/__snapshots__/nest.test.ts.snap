// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nestSerializer works correctly 1`] = `"namespace Address {namespace Japan {namespace Tokyo {type Minato = { inTokyo: true; endWithA: false };type Chiyoda = { inTokyo: true; endWithA: true };}namespace Osaka {type Umeda = { inTokyo: false; endWithA: true };}}}"`;

exports[`nestSerializer works correctly 2`] = `
"namespace Address {
  namespace Japan {
    namespace Tokyo {
      type Minato = { inTokyo: true; endWithA: false };
      type Chiyoda = { inTokyo: true; endWithA: true };
    }
    namespace Osaka {
      type Umeda = { inTokyo: false; endWithA: true };
    }
  }
}
"
`;
